### ISSUE TITLE
Regression: ICE: no kind for cast

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#37018]
A [stable-to-beta crater report](https://gist.github.com/alexcrichton/ab516d4d73470c9da0884780d716fd2b) shows regressions related to 

```
error: internal compiler error(B: ../src/librustc_passes/consts.rs:483: no kind for cast(B
    (B--> (B/home/crate/target/debug/build/gl-a556014da90362fa/out/bindings.rs:5850:20(B
     (B|(B
5850(B (B| (B                f: (Bsuper::missing_fn_panic as *const libc::c_void(B,(B
     (B| (B                   (B^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^(B

```

* [flywheel-0.1.1](https://crates.io/crates/flywheel) ([before](https://tools.taskcluster.net/task-inspector/#u0eTxWe4RxGCxrH6JKueOw)) ([after](https://tools.taskcluster.net/task-inspector/#PDGyNchmSkue47qiYsr_RA))
* [hematite-client-0.0.3](https://crates.io/crates/hematite-client) ([before](https://tools.taskcluster.net/task-inspector/#8xNhufbFSH6v0iuFmbYF4A)) ([after](https://tools.taskcluster.net/task-inspector/#w2u5QaTfQ1S831IU_MVoVA))

I'm opening an issue to ensure that we're tracking these regressions and they're either expected (so we can close), this is a dupe (so we can close), or we can fix (and then close!)

cc @rust-lang/compiler

**Comment by TimNN** [comment]
Minimized:

```rust
mod foo {
    extern crate libc;
}

use foo::libc;

pub static BAR: *const libc::c_void = baz as *const libc::c_void;

fn baz() { loop {} }

fn main() {}
```

Backtrace:

```
error: internal compiler error: ../src/librustc_passes/consts.rs:483: no kind for cast
 --> regice.rs:7:39
  |
7 | pub static BAR: *const libc::c_void = baz as *const libc::c_void;
  |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', ../src/librustc_errors/lib.rs:602
stack backtrace:
   1:        0x111422288 - std::sys::backtrace::tracing::imp::write::h15f4dcb9fc5cc76b
   2:        0x11142f9ff - std::panicking::default_hook::{{closure}}::h122a1c9dfacef96f
   3:        0x11142cefd - std::panicking::default_hook::h9cf83dcedc626223
   4:        0x11142d596 - std::panicking::rust_panic_with_hook::hc99436623a03228e
   5:        0x10e2cbefb - std::panicking::begin_panic::h17785a79ba05422e
   6:        0x10e2e8473 - rustc::session::opt_span_bug_fmt::{{closure}}::h213520e1e2445968
   7:        0x10e2d7789 - rustc::session::span_bug_fmt::h887f1af1528af01b
   8:        0x10e2dfff6 - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'v>>::visit_expr::h0be89430a25e80da
   9:        0x10e2ea346 - rustc_passes::consts::CheckCrateVisitor::global_expr::{{closure}}::h76e0a472abd5b020
  10:        0x10e2dde77 - rustc_passes::consts::CheckCrateVisitor::global_expr::h170e19160d2d6eb6
  11:        0x10e2e0f61 - rustc_passes::consts::check_crate::h91cf5a98b4344a74
  12:        0x10d87c814 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::h5dabeea7050cf7c2
  13:        0x10d8530b6 - rustc_driver::driver::phase_3_run_analysis_passes::h79f20d1127b1dfba
  14:        0x10d840eca - rustc_driver::driver::compile_input::h8ef818979445c42c
  15:        0x10d868fa8 - rustc_driver::run_compiler::ha981cef7568b5eb8
  16:        0x10d7a9700 - std::panicking::try::do_call::h4a286c81408a0d92
  17:        0x11142ffba - __rust_maybe_catch_panic
  18:        0x10d7c91c4 - <F as alloc::boxed::FnBox<A>>::call_box::h5de1378b9690a33a
  19:        0x11142c1b4 - std::sys::thread::Thread::new::thread_start::h2eacc7b7061d5d8c
  20:     0x7fff9cd4b99c - _pthread_body
  21:     0x7fff9cd4b919 - _pthread_start
```

**Comment by TimNN** [comment]
Introduced between `nightly-2016-08-21` and `nightly-2016-08-22` ([Changes](https://github.com/rust-lang/rust/compare/490189634...1576de0ce)),

This was also caused by #35776, could be a duplicate of #36747, and thus may as well be fixed by #37023.

cc @jseyfried

**Comment by nikomatsakis** [comment]
@TimNN nice hunting! :)

Closing as duplicate of #36747
